nodejs/week2/app.js

```diff
@@ -0,0 +1,87 @@
+const express = require("express");
+const app = express();
+const port = process.env.PORT || 3000;
+const documents = require("./documents");
+
+app.use(express.json());
+app.use(express.urlencoded({ extended: true }));
+
+app.get("/search", (req, res) => {
+  try {
+    const query = req.query.q;
+    if (!query) {
+      res.status(200).json(documents);
+    } else if (!isNaN(+query)) {
+      const result = documents.filter((e) => e.id === +query);
+      result.length === 0
+        ? res.status(200).json(documents)
+        : res.status(200).json(result);
+    } else {
+      const result = documents.filter((e) =>
+        Object.values(e).some(
+          (v) =>
+            typeof v === "string" &&
+            v.toLowerCase().includes(query.toLowerCase())
+        )
+      );
+      result.length === 0
+        ? res.status(200).json(documents)
+        : res.status(200).json(result);
+    }
+  } catch (err) {
+    res.status(500).json(err);
+  }
+});
+app.get("/documents/:id", (req, res) => {
+  try {
+    const id = req.params.id;
+    const result = documents.filter((e) => e.id === +id);
+    result.length === 0
+      ? res.status(404).json("no result with this id")
+      : res.status(200).json(result);
+  } catch (err) {
+    res.status(500).json(err);
+  }
+});
+app.post("/search", (req, res) => {
+  try {
+    const query = req.query.q;
+    const fields = req.body.fields;
+    if (!query && !fields) {
+      res.status(200).json(documents);
+    } else if (!query && fields) {
+      const result = document.filter((e) =>
+        Object.keys(fields).every((k) => fields[k] === e[k])
+      );
+      result.length === 0
+        ? res.status(404).json("no result with this id")
+        : res.status(200).json(result);
+    } else if (query && !fields) {
+      if (!isNaN(+query)) {
+        const result = documents.filter((e) => e.id === +query);
+        result.length === 0
+          ? res.status(200).json("no result found")
+          : res.status(200).json(result);
+      }
```

Comment:
Well, there is a bug which you can fix. The problem is if a user send a request q=text, so what will happen. Just nothing, the user is not going to receive any response. because the first condition is met, but the second condition didn't. So the machine doesn't which line of code it should execute.
**Visualisation:**
`
else if(query && !fields{    // this condition met because the query has value
   if (!isNaN(+query)) {       // this condition didn't met because the value is not a number
    }
  // so machine is stuck here because there no else condition or code to execute
} `